### PR TITLE
Implementation of WFJT limit & SCM_branch prompting

### PR DESCRIPTION
**Pull request by AlanCoding**
##### SUMMARY
Connect https://github.com/ansible/awx/issues/1731

Does not yet do https://github.com/ansible/awx/issues/1845

There will also be a need to apply this to https://github.com/ansible/awx/issues/282

I believe that 282 will be in scope near-term, but 1845 will not.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - API

##### AWX VERSION
```
6.0.0
```

##### ADDITIONAL INFORMATION
Heads up to @jbradberry, this is about as simple as this material is going to get, and that's not very simple. This does the abuse of Django character fields to make them nullable, and tries to make it so that `scm_revision` will follow suit in a relatively consistent pattern.

### CODE REVIEW

**Comment by jakemcdermott** [comment]
recheck

**Comment by AlanCoding** [comment]
This implementation is still a half-way move toward making WFJTs full "config" objects. Basically, we have the opportunity to remove some of the really back hacks that had to be implemented for sliced jobs. See:

https://github.com/ansible/awx/blob/76a10991ae4a963286ec6042a2d8758b3ca2272e/awx/main/models/workflow.py#L232-L240

I'm resisting doing a much broader refactor to get rid of this. Ideally this would use `self.workflow_job.prompts_dict()` instead of the config object. But it still wouldn't be clean, because of the exception for variable acceptance with extra_vars.

**Review by mabashian — DISMISSED** [review]
This looks great - I did find one scenario that's not covered in the UI currently.  If you create a workflow and prompt for limit on launch (make sure this is the only promptable thing on the workflow) and launch that workflow, the prompt modal isn't shown and the job is simply launched.  In order to fix this I think we need to change https://github.com/ansible/awx/blob/devel/awx/ui/client/lib/models/WorkflowJobTemplate.js#L57 to

```
    return (
        launchData.can_start_without_user_input &&
        !launchData.ask_inventory_on_launch &&
        !launchData.ask_variables_on_launch &&
        !launchData.survey_enabled &&
        !launchData.ask_scm_branch_on_launch &&
        !launchData.ask_limit_on_launch &&
        launchData.variables_needed_to_start.length === 0
    );
```

Note the addition of `!launchData.ask_limit_on_launch &&` to that conditional.  Otherwise I think this is looking good.  I tested out the workflow node scenarios and ran into https://github.com/ansible/awx/issues/4293 but that should be addressed by https://github.com/ansible/awx/pull/4525

**Comment by AlanCoding** [comment]
Thanks, good catch!

I was easily able to replicate this situation manually and confirmed that the latest commit fixes it.

While doing this, I noticed another odd thing - I couldn't set it to prompt for limit while creating the WFJT. That turned out to be because it was missing in yet another place. Also confirmed fixed in last commit.

**Review by mabashian — APPROVED** [review]
lgtm

**Review by jlmitch5 — APPROVED** [review]
both prompting for limit on WFJT is available, and makes the modal pop up when you launch it.  Looks good!

File: awx/api/serializers.py [hunk 1]
```diff
@@ -3344,6 +3347,22 @@ def get_related(self, obj):
...
+        attrs = super(WorkflowJobTemplateSerializer, self).validate(attrs)
+
+        # process char_prompts, these are not direct fields on the model
+        mock_obj = self.Meta.model()
+        for field_name in ('scm_branch', 'limit'):
+            if field_name in attrs:
+                setattr(mock_obj, field_name, attrs[field_name])
+                attrs.pop(field_name)
+
+        # Model `.save` needs the container dict, not the psuedo fields
```
Comment (jbradberry) [inline awx/api/serializers.py]: pseudo

File: awx/main/utils/common.py [hunk 2]
```diff
@@ -435,16 +440,50 @@ def model_to_dict(obj, serializer_mapping=None):
...
 
+class CharPromptDescriptor:
+    """Class used for identifying nullable launch config fields from class
+    ex. Schedule.limit
+    """
+    def __init__(self, field):
+        self.field = field
+
+
+class NullablePromptPsuedoField:
```
Comment (jbradberry) [inline awx/main/utils/common.py]: Can we fix the name here?

File: awx/main/utils/common.py [hunk 3]
```diff
@@ -454,15 +493,29 @@ def copy_model_by_class(obj1, Class2, fields, kwargs):
...
+                if field_name in obj1.char_prompts:
+                    new_val = obj1.char_prompts[field_name]
+            elif hasattr(obj1, field_name):
+                # extremely rare case where a template spawns a launch config - sliced jobs
+                new_val = getattr(obj1, field_name)
+            if new_val is not None:
+                create_kwargs.setdefault('char_prompts', {})
+                create_kwargs['char_prompts'][field_name] = new_val
+        elif isinstance(descriptor, ManyToManyDescriptor):
+            continue  # not coppied in this method
```
Comment (jbradberry) [inline awx/main/utils/common.py]: copied

**Comment by dsesami** [comment]
While we're here: Can we update the tooltip for the limit field in workflows? It's the inventory limit, but that isn't clarified for the user: 

Screenshot from 2019-09-09 14-20-16

A bit nitpicky, but would likely help users. Maybe even a docs link. cc @tvo318 

In addition, the prompt tooltip actually provides better information:
Screenshot from 2019-09-09 14-23-09

update: @jakemcdermott is making changes here. Once he does so, I'll give a final pass-through. Looks good so far.

File: awx/main/utils/common.py (hunk 2 above)
Comment (AlanCoding) [reply awx/main/utils/common.py]: Oh, you just mean `Psuedo` -> `Pseudo`

Yeah, of course. Thanks for finding all of these :-P

**Review by jakemcdermott — APPROVED** [review]

**Review by dsesami — APPROVED** [review]
Looking really solid. Pinging @squidboylan on a final call for this on QE side, looks good from UI end.

**Review by jbradberry — APPROVED** [review]

**Review by squidboylan — APPROVED** [review]

**Comment by AlanCoding** [comment]
recheck

**Pull request merged**